llm_benchmark.py

```diff
@@ -401,31 +406,42 @@ def make_gemini_messages(prompt: str, files: List[InputFile]):
 
 async def gemini_chat(context: ApiContext) -> ApiResult:
     async def chunk_gen(response) -> TokenGenerator:
+        tokens = 0
         async for chunk in make_json_chunk_gen(response):
-            content = chunk["candidates"][0].get("content", None)
+            content = chunk["candidates"][0].get("content")
             if content and "parts" in content:
                 part = content["parts"][0]
                 if "text" in part:
+                    tokens += 1
                     yield part["text"]
+            usage_metadata = chunk.get("usageMetadata")
+            if usage_metadata:
+                num_tokens = usage_metadata.get("candidatesTokenCount")
+                while tokens < num_tokens:
+                    tokens += 1
+                    yield ""
 
-    url = f"https://generativelanguage.googleapis.com/v1beta/models/{context.model}:streamGenerateContent?key={get_api_key('GOOGLE_GEMINI_API_KEY')}"
-    headers = make_headers()
-    harm_categories = [
-        "HARM_CATEGORY_HARASSMENT",
-        "HARM_CATEGORY_HATE_SPEECH",
-        "HARM_CATEGORY_SEXUALLY_EXPLICIT",
-        "HARM_CATEGORY_DANGEROUS_CONTENT",
-    ]
+    # The Google AI Gemini API (URL below) doesn't return the number of generated tokens.
+    # Instead we use the Google Cloud Vertex AI Gemini API, which does return the number of tokens, but requires an Oauth credential.
+    # Also, setting safetySettings to BLOCK_NONE is not supported in the Vertex AI Gemini API, at least for now.
+    # url = f"https://generativelanguage.googleapis.com/v1beta/models/{context.model}:streamGenerateContent?key={get_api_key('GOOGLE_GEMINI_API_KEY')}"
+    url, headers = make_google_url_and_headers(args.model, "streamGenerateContent")    
+    # harm_categories = [

```

Comment:
Is this worth keeping around?

Reply:
I'm waiting to hear back from GCP support on how exactly to get the behavior we want from Gemini, once I get an answer there I'll rework this section accordingly. However the team responsible for this filtering has been "running around with their hair on fire", so it might be a week or two before we can resolve this.